Feature/composite lennard jones

yangmr04 opened the pull request:
# Breaking changes
- New `LennardJonesParametersSets` supertype which has `LJParameters` and `CompositeLJParameters{C}` as subtypes, where `C` is the number of components of the system.
- Updated `AtomWalker{C}` to have a type parameter `C`, the number of components.
- Updated the way how inter-component and intra-component energies are evaluated.
- Updated `AtomsMCMoves` module to accommodate the changes above.
- Updated `FreeBirdIO` module to accommodate the changes above.

# Other changes
- Added several convenient functions such as `split_components_by_chemical_species` and `sort_components_by_atomic_number`. 
- Split the above mentioned functions into a `helpers.jl` under `EnergyEval`.
- Updated the example script to work with the overhauled system.
- Added several `show` methods to make the repl print out look nicer.

junchichen21 commented on Project.toml:
```diff
@@ -21,6 +21,7 @@ Random = "9a3f8284-a2c9-5f02-9a11-845980a1fd5c"
 Reexport = "189a3867-3050-52da-a836-e630ba90ab69"
 Setfield = "efcf1570-3423-57d1-acb7-fd33fddbac46"
 StaticArrays = "90137ffa-7385-5640-81b9-e52037218182"
+Statistics = "10745b16-79ce-11e8-11f9-7d13ad32a3b2"
```
Why do we need this package?

yangmr04 replied:
It's needed for some convenient stats functions in module `AbstractWalkers`. `Statistics` is a base Julia package, shouldn't add any load to the performance.

rwexler commented on scripts/example.jl:
```diff
@@ -2,26 +2,27 @@
 using FreeBird
 
 # First, we need to read the walkers from a file
+# Note that the periodic boundary conditions are updated to True, True, False
```
Does "False" mean they can have any z value or they're confined, by an elastically reflecting wall, to a container?

yangmr04 replied:
Confined by a reflecting wall.

rwexler commented on scripts/example.jl:
```diff
@@ -2,26 +2,27 @@
 ats = read_walkers("scripts/slab-08-4x4.extxyz",pbc="TTF")
-# Note that the periodic boundary conditions are set to True, True, False
+
+# Update the system to have 80 frozen particles and 8 free particles
```
This is based on knowledge of the structure that was read in, correct?

yangmr04 replied:
Yes, the xyz file contains the same info in this case, but it can be overwritten by the keyword argument `pbc="TTF"` if modification of the boundary conditions is needed.

rwexler commented on scripts/example.jl:
```diff
@@ -2,26 +2,27 @@
+ls = LJAtomWalkers(ats, lj)
 
 # We can now create a Monte Carlo object
-mc = MCDemonWalk(max_add_steps=10000, e_demon_tolerance=0.01u"eV")
```
Do we have any tests of MC random vs. demon walks?

yangmr04 replied:
Not yet. By solely doing demon walks, we can't make the NS energies to go down consistently, i.e., NS will stuck at the lowest energy from the initial set of walkers. A mixed strategy is available but is slower than pure MC random walks. Will continue working on this.

yangmr04 replied:
Demon walk should be included. It works, just need some more tests/benchmarks to show how well (or bad) it works.

rwexler commented on scripts/example.jl:
```diff
@@ -2,26 +2,27 @@
 
 # And finally, we can run the simulation
-energies, ls, _ = nested_sampling_loop!(ls, 10000, mix, save)
+energies, ls, _ = nested_sampling_loop!(ls, ns_params, 10_000, mc, save)
```
Is 10,000 the number of nested sampling iterations?

yangmr04 replied:
Yes. The number of decorrelation steps is defined within `ns_params`.

rwexler commented on src/AbstractWalkers/AtomWalkers.jl:
```diff
@@ -9,79 +9,187 @@ The `AtomWalker` struct represents a walker composed of atoms/molecules.
 
 """
-mutable struct AtomWalker 
+mutable struct AtomWalker{C}
```
What does the {C} mean?

yangmr04 replied:
It means `AtomWalker` is a parametrized type, where is this case, `C` is the number of components.

rwexler commented on src/Potentials.jl:
```diff
@@ -115,6 +120,84 @@ end
+function CompositeLJParameters(c::Int, ljs::Vector{LJParameters})
+    if length(ljs) == c^2
+        # If the number of LJParameters sets is equal to the number 
+        # of elements in the matrix, assuming that Vector{LJParameters} 
```
```suggestion
        # of elements in the matrix, then assume that Vector{LJParameters} 
```

rwexler commented on src/Potentials.jl:
```diff
@@ -115,6 +120,84 @@ end
+    if length(ljs) == c^2
+        # If the number of LJParameters sets is equal to the number 
+        # of elements in the matrix, assuming that Vector{LJParameters} 
+        # is a flattened matrix, then reshape the vector into a matrix.
```
```suggestion
        # is a flattened matrix and reshape the vector into a matrix.
```

rwexler commented on src/Potentials.jl:
```diff
@@ -115,6 +120,84 @@ end
+    else
+        throw(ArgumentError("the number of LJParameters sets is not compatible with the number of components."))
+    end
+end
```
this dispact requires a tame user; should these functions spit out a message saying something like, "by specifying X sets of LJ parameters, <this function> has assumed <assumption>. If this was not your intention, please check the documentation or raise an issue."

yangmr04 replied:
Good idea! Done!

yangmr04 commented on src/AbstractWalkers/AtomWalkers.jl:
````diff
@@ -9,79 +9,187 @@ The `AtomWalker` struct represents a walker composed of atoms/molecules.
+```
+
 """
-function assign_lj_energies!(walkers::Vector{AtomWalker}, lj::LJParameters)#; frozen::Int64=0, e_frozen=0.0u"eV")
````
It was replaced by `assign_energy!`:
https://github.com/wexlergroup/FreeBird.jl/blob/ae01f26765487b6733c745b03b3729d495d1b0b2/src/AbstractWalkers/AtomWalkers.jl#L142

yangmr04 commented on src/AbstractWalkers/AtomWalkers.jl:
```diff
@@ -9,79 +9,187 @@ The `AtomWalker` struct represents a walker composed of atoms/molecules.
+        .--------------.  
+       /|Fel           |  
+      / H   H   Cl     |  
+     /  Hu   O         |  
```
That's true. It's a design by `AtomsBase`. It supposes to give a quick look of the structure and not meant to be comprehensive.

yangmr04 commented on src/EnergyEval/EnergyEval.jl:
```diff
@@ -0,0 +1,294 @@
+                       ljs::CompositeLJParameters{C}, 
+                       list_num_par::Vector{Int},
+                       frozen::Vector{Bool}
+                       ) where {C}
```
Good point, we should...

yangmr04 replied:
Fixed! Let's have a new line to end the call signatures.

yangmr04 commented on src/EnergyEval/EnergyEval.jl:
Everything changed :D The old functions such as `free_free_energy()` and `free_frozen_energy()` were replaced by universal `inter_component_energy()` and `intra_component_energy()` functions.

yangmr04 commented on src/FreeBirdIO.jl:
```diff
@@ -64,10 +64,12 @@ function convert_system_to_walker(at::FlexibleSystem, resume::Bool)
     energy = (haskey(data, :energy) && resume) ? data[:energy]*u"eV" : 0.0u"eV"
     iter = (haskey(data, :iter) && resume && data[:iter]>=0) ? data[:iter] : 0
-    num = (haskey(data, :num_frozen_part)  && resume) ? data[:num_frozen_part] : 0
+    list_num_par = (haskey(data, :list_num_par) && resume) ? data[:list_num_par] : [length(at)]
```
When reading in an .extxyz file, if the file already contains these "keys" and we want to resume the calculation, they will be restored. Otherwise, each of them gets reassigned a new starting value.

yangmr04 replied:
For example:
```julia
energy = (haskey(data, :energy) && resume) ? data[:energy]*u"eV" : 0.0u"eV"
```
is equivalent to:
```julia
if haskey(data, :energy) && resume # both true
    energy = data[:energy]*u"eV" # read in the data
else
    energy = 0.0u"eV" # assign 0 eV as the new starting energy
end
```

yangmr04 commented on src/FreeBirdIO.jl:
```diff
@@ -196,20 +198,23 @@ function convert_walker_to_system(at::AtomWalker)
+function append_walker(filename::String, at::AtomWalker{C}) where C
+    # ats = read_walkers(filename)
+    sys = convert_walker_to_system(at)
+    # push!(ats, at)
```
Wasn't sure the version was working, kept them around as backup. I think we can now delete them.

yangmr04 commented on src/Potentials.jl:
```diff
@@ -115,6 +120,84 @@ end
+CompositeLJParameters is a struct that represents a set of composite Lennard-Jones parameters.
+
+# Fields
+- `lj_param_sets::Matrix{LJParameters}`: A matrix of LJParameters representing the LJ parameter sets.
```
Done!

yangmr04 commented on src/AtomsMCMoves.jl:
```diff
@@ -63,21 +63,21 @@ end
 
 """
-function mean_sq_displacement(at::AtomWalker, at_orig::AtomWalker)
+function mean_sq_displacement(at::AtomWalker{C}, at_orig::AtomWalker{C}) where C
```
When the function doesn't need to know the value of `C`. I.e. this function should work for any number of components.

yangmr04 commented on src/AbstractWalkers/AtomWalkers.jl:
```diff
@@ -9,79 +9,187 @@ The `AtomWalker` struct represents a walker composed of atoms/molecules.
-- `walkers::Vector{AtomWalker}`: The list of `AtomWalkers`.
-- `lj_potential::LJParameters`: The Lennard-Jones potential parameters.
+- `walkers::Vector{AtomWalker{C}}`: A vector of atom walkers, where `C` is the number of components.
+- `lj_potential::LennardJonesParametersSets`: The Lennard-Jones potential parameters. See `LennardJonesParametersSets`.
```
It's kind of carried over from `LJParameters` and attached with `Sets`. I suggest we change it later, I sense it will be some changes around how we handle LJ potentials (or import it from another performant package).

yangmr04 commented on scripts/example.jl:
```diff
@@ -2,26 +2,27 @@
 
-# We can also create a save object
-save = SaveEveryN("output-energies.csv",100)
+ns_params = NestedSamplingParameters(1600, 0.1, 0.01, 1e-5, 0.1, 0, 200)
```
Done!

yangmr04 commented on src/AbstractWalkers/AtomWalkers.jl:
```diff
@@ -9,79 +9,187 @@ The `AtomWalker` struct represents a walker composed of atoms/molecules.
+    frozen = zeros(Bool, C)
+    if !isempty(freeze_species)
+        elements = unique(atomic_symbol(configuration))
+        @show elements
```
Similar to `println()` but also shows the variable name, not just the values.

yangmr04 commented on src/AtomsMCMoves.jl:
```diff
@@ -101,6 +101,17 @@ function single_atom_random_walk!(pos::SVector{3,T}, step_size::Float64) where T
     return pos 
 end
 
+function free_par_index(at::AtomWalker{C}) where C
```
Added. It's an internal function, not exported.

yangmr04 commented on src/EnergyEval.jl:
It respawned into a subdir 😂

yangmr04 commented on src/EnergyEval/helpers.jl:
Done!

yangmr04 commented:
@rwexler I have addressed all your comments, many thanks for the suggestions, they are great! I think it's now ready to merge. @junchichen21 @D-Thibodeaux Feel free to take another look before I merge the pull request.
